### ISSUE TITLE
Discuss guidelines for the implementation of algorithms with randomness

### ISSUE THREAD

**Issue by salvisolamartinell** [bsc-wdc/dislib#100]
The sequential approach of reproducible randomness is using a single RandomState instance. If not provided, numpy random functions use a default RandomState (I don't know how it's created). On a distributed system, this approach is not possible because passing the same RandomState instance to multiple tasks will produce the same numbers. We have to find a method that guarantees the following:

1. If seed or RandomState is provided to the application, the result must be reproducible.
2. Avoid different tasks producing the same numbers (I have observed this problem in local using numpy.random.randint(): it looked like numpy's default RandomState was the same for multiple tasks).
3. Avoid hanging programs due to lack of entropy (I'm not sure which was the problem, but I have had programs in local hanging until I clicked somewhere).
4. Ideally, if no seed nor RandomState are provided, the number of possible results of a task should be as big as the number of possible RandomState states.

**Comment by salvisolamartinell** [comment]
The approach I thought:
If seed or RandomState is provided to the application: use this RandomState or RandomState(seed), if no RanomState is provided use np.random.mtrand._rand (like in sklearn.utils.check_random_state). Then pass a new generated seed (from 0 to 2^32-1) to each task. In the task, create np.random.RandomState(task_seed) and use it.

Tasks with randomness always use seeds, not numpy.random methods! (otherwise we may have duplicate random generation).

Complies with points 1, 2 and 3. The number of possible results for each task is 2^32. Expected first collision after about 2^16 (65000) tasks. To comply with point 4, we would need to call RandomState.set_state() in the task and set it using the seed and the time().

**Comment by salvisolamartinell** [comment]
I have tried to test 2 methods:
1. using numpy.random for each task
2. passing generated RandomState instances to the tasks, as recommended by @javicid 

I cannot reproduce the problems I mentioned previously with approach 1. Both approaches seem to work well now, both in local and MN4.

Code tried:
```
import numpy as np
from pycompss.api.api import compss_wait_on
from pycompss.api.task import task


@task(returns=1)
def generate_int_default():
    return np.random.randint(0, 2**32-1)


@task(returns=1)
def generate_int_given(random_state):
    return random_state.randint(0, 2 ** 32 - 1)


def main():
    n_ints_to_generate = 100
    ints_default = []
    ints_given = []
    print('here')
    for i in range(n_ints_to_generate):
        ints_default.append(generate_int_default())
    for i in range(n_ints_to_generate):
        ints_given.append(generate_int_given(np.random.RandomState()))
    ints_default = compss_wait_on(ints_default)
    ints_given = compss_wait_on(ints_given)

    print(ints_default)
    print(ints_given)

    print('Number of ints to generate for each case:', n_ints_to_generate)
    print('Number of unique ints generated with default random states: ',
          len(set(ints_default)))
    print('Number of unique ints generated with given random states: ',
          len(set(ints_given)))


if __name__ == "__main__":
    main()
```
Results:
```
Number of ints to generate for each case: 100
Number of unique ints generated with default random states:  100
Number of unique ints generated with given random states:  100
```

So I think we can follow method 2 (also with reproducibility), explain it in the guidelines and close this issue.

**Comment by javicid** [comment]
As far as I understand, this means that tasks generate random numbers correctly and we do not need to care about passing random_state unless specified by the user?

What happens if you pass the same RandomState instance to multiple tasks?

**Comment by salvisolamartinell** [comment]
>As far as I understand, this means that tasks generate random numbers correctly and we do not need to care about passing random_state unless specified by the user?
No, we haven't observed any repetition but they could occur on a different execution/machine/implementation. numpy.random is not known to be thread-safe, so we shouldn't use it in tasks. We have to pass a state to tasks or use python's random module (which is thread-safe but slower).
> What happens if you pass the same RandomState instance to multiple tasks?
All tasks generate the same int (tried in local and mn4).

**Comment by javicid** [comment]
>  numpy.random is not known to be thread-safe

Each task runs in a different process, so this shouldn't be a problem.

**Comment by salvisolamartinell** [comment]
It could be a problem with multi-processing too:
https://stackoverflow.com/questions/12915177/same-output-in-different-workers-in-multiprocessing

**Comment by javicid** [comment]
I'm not sure if the multiprocessing module really works with multiple processes as we do. PyCOMPSs runs each task using a different python process started manually.

**Comment by kafkasl** [comment]
PyCOMPSs runs each task inside a multiprocessing.Process(). So they are different processes but it's handled by multiprocessing no manually.

**Comment by javicid** [comment]
Thanks for the correction. 

We have decided to pass a different seed to each task. This means that with 65k tasks the chance for repeated behavior is 40%, which is something we can live with, so I'm closing the issue.
